File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/presenter/HttpCallListPresenter.java
package com.prateekj.snooper.networksnooper.presenter;

import com.prateekj.snooper.networksnooper.adapter.HttpCallListClickListener;
import com.prateekj.snooper.networksnooper.database.SnooperRepo;
import com.prateekj.snooper.networksnooper.model.HttpCallRecord;
import com.prateekj.snooper.networksnooper.views.HttpListView;

import java.util.List;

import static com.prateekj.snooper.utils.CollectionUtilities.last;

public class HttpCallListPresenter implements HttpCallListClickListener {
  public static final int PAGE_SIZE = 20;
  private final SnooperRepo snooperRepo;
  private HttpListView httpListView;
  private long lastCallId;

  public HttpCallListPresenter(HttpListView httpListView, SnooperRepo snooperRepo) {
    this.httpListView = httpListView;
    this.snooperRepo = snooperRepo;
  }

  public void init() {
    List<HttpCallRecord> httpCallRecords = snooperRepo.findAllSortByDateAfter(-1, PAGE_SIZE);
    lastCallId = last(httpCallRecords).getId();
    httpListView.initHttpCallRecordList(httpCallRecords);
  }

  public void onNextPageCall() {
    List<HttpCallRecord> httpCallRecords = snooperRepo.findAllSortByDateAfter(lastCallId, PAGE_SIZE);
    lastCallId = last(httpCallRecords).getId();
    httpListView.appendRecordList(httpCallRecords);
  }

  @Override
  public void onClick(HttpCallRecord httpCall) {
    httpListView.navigateToResponseBody(httpCall.getId());
  }

  public void onDoneClick() {
    httpListView.finishView();
  }

  public void onDeleteRecordsClicked() {
    httpListView.showDeleteConfirmationDialog();
  }

  public void confirmDeleteRecords() {
    snooperRepo.deleteAll();
    httpListView.updateListViewAfterDelete();
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/presenter/HttpCallListPresenter.kt
package com.prateekj.snooper.networksnooper.presenter

import com.prateekj.snooper.networksnooper.adapter.HttpCallListClickListener
import com.prateekj.snooper.networksnooper.database.SnooperRepo
import com.prateekj.snooper.networksnooper.model.HttpCallRecord
import com.prateekj.snooper.networksnooper.views.HttpListView

class HttpCallListPresenter(
  private val httpListView: HttpListView,
  private val snooperRepo: SnooperRepo
) : HttpCallListClickListener {
  private var lastCallId: Long = 0

  fun init() {
    val httpCallRecords = snooperRepo.findAllSortByDateAfter(-1, PAGE_SIZE)
    if (httpCallRecords.isEmpty()) {
      httpListView.renderNoCallsFoundView()
      return
    }
    lastCallId = httpCallRecords.last().id
    httpListView.initHttpCallRecordList(httpCallRecords)
  }

  fun onNextPageCall() {
    val httpCallRecords = snooperRepo.findAllSortByDateAfter(lastCallId, PAGE_SIZE)
    lastCallId = httpCallRecords.last().id
    httpListView.appendRecordList(httpCallRecords)
  }

  override fun onClick(httpCall: HttpCallRecord) {
    httpListView.navigateToResponseBody(httpCall.id)
  }

  fun onDoneClick() {
    httpListView.finishView()
  }

  fun onDeleteRecordsClicked() {
    httpListView.showDeleteConfirmationDialog()
  }

  fun confirmDeleteRecords() {
    snooperRepo.deleteAll()
    httpListView.updateListViewAfterDelete()
  }

  companion object {
    const val PAGE_SIZE = 20
  }
}
